internal/engine/upper_test.go

```diff
@@ -272,16 +283,41 @@ func (b *fakeBuildAndDeployer) BuildAndDeploy(ctx context.Context, st store.RSto
 		b.resultsByID[key] = val
 	}
 
+	// block until we know we're supposed to resolve this build
+	for {
+		if _, ok := b.buildsAwaitingResolution[count]; ok {
+			delete(b.buildsAwaitingResolution, count)
+			break
+		}
+
+		select {
+		case <-ctx.Done():
+			break
+		case <-time.After(time.Millisecond * 5):
+		}
+	}
+
 	return result, err
 }
 
 func newFakeBuildAndDeployer(t *testing.T) *fakeBuildAndDeployer {
 	return &fakeBuildAndDeployer{
-		t:              t,
-		calls:          make(chan buildAndDeployCall, 20),
-		buildLogOutput: make(map[model.TargetID]string),
-		resultsByID:    store.BuildResultSet{},
+		t:                        t,
+		calls:                    make(chan buildAndDeployCall, 20),
+		buildsAwaitingResolution: make(map[int]bool),
+		buildLogOutput:           make(map[model.TargetID]string),
+		resultsByID:              store.BuildResultSet{},
+	}
+}
+
+func (b *fakeBuildAndDeployer) resolveBuildAtIndex(i int) {
```

Comment:
could rename the arg to `index` and drop `AtIndex` from the function name, if you wanted